### PR TITLE
[BUG FIX] Fix invalid gradients in differentiable rigid-body simulation.

### PR DESCRIPTION
**Pull request by duburcqa**
## Description

Follow-up to #2842, fixing the gradient defects surfaced by a full audit of the differentiable rigid-body backward pass, plus the maintainability and convention issues found along the way.

Gradient fixes:
- Checkpoint restore (`RigidSolver.load_ckpt`) reran the cartesian-space update but not the forward velocity pass, while the replayed substeps skip both; every backward primal in the window then carried stale link velocities, injecting an eps-independent error into every multi-step gradient.
- The equality constraint reverse dropped the velocity-product bias `Jdot @ qvel` folded into `aref` (CONNECT and WELD position rows) and the WELD rotation rows' quaternion-derivative bias.
- The constraint reverses accumulated `d(jac_qvel)/d(vel)` once per chain visit, double-counting shared ancestor dofs of same-root pairs.
- The collision constraint reverse ignored `impratio` in the regularized friction-cone coefficient of `efc_D`.
- The prismatic forward-kinematics reverse routed the link position gradient into the joint-frame quaternion term, corrupting quaternion gradients of chains with an offset slide joint (2e-2 relative error).
- `dofs.pos.grad` (position control, joint stiffness) was never consumed; it now folds back into `qpos.grad` for all joint types.
- Fixed root links' pose gradients were zeroed by the forward-kinematics reverse even though their pose is persistent input state.
- The adjoint conjugate-gradient solve of the constraint backward stopped on an absolute residual threshold and clamped its step coefficients at an absolute epsilon, capping gradient accuracy around 1e-8 relative whenever the upstream gradient is small; it now converges to the round-off floor.
- The input tape keyed tracked setters by method name only, so two same-step calls on distinct dof subsets (arm + gripper) evicted each other. The key now includes the dof subset, canonicalized for slice (only hashable from Python 3.12) and device-tensor indices, and the reverse skips scalar / array-like inputs; all are valid setter inputs that crashed the backward pass.
- Unsupported configurations now raise at build time instead of silently computing wrong gradients: torsional/rolling friction and `attach()` with `requires_grad=True`.
- Collision-row count in the joint-limit reverse follows `rows_per_contact` instead of a hardcoded 4.

Maintainability: the duplicated impedance-derivative, velocity-contraction, and plane-contact-frame blocks collapse into shared funcs; dead code is removed; grad shuffles on the backward hot path go through zero-copy torch views with the kernels as fallback.

Tests: existing tests are strengthened to fail on each defect above (same-root four-bar connect and nested skew-axis weld fixtures, offset-slide chain, joint stiffness, non-default `impratio`, split-subset force control, larger equality velocities), and the suite is aligned with the testing conventions. All finite-difference tolerances are re-pinned to floors measured on darwin cpu/Metal and linux cpu/CUDA after the fixes, at a x1.5-5 margin over the worst platform: most fp64 checks tighten by 1 to 5 orders of magnitude (the old floors were these bugs, not physics). Two multistep arms whose position loss was constant now read the quaternion instead of passing vacuously.

## Related Issue

Follow-up to #2842.
Addresses #2069: the control-force gradient path it reports broken is fixed and validated per-step by the strengthened multistep tests.
Addresses #2150: the forward-kinematics backward pass it exercises is audited and fixed here; the specific Franka scene remains to be confirmed.

## Motivation and Context

The audit found gradient errors ranging from 2e-2 relative down to 3e-7 that finite-difference validation had missed because the scenarios or tolerances did not excite them. Wrong gradients silently break any downstream optimization.

## How Has This Been / Can This Be Tested?

- `pytest tests/grad/ --backend cpu`: 46 passed, 10 skipped (Apple Metal reverse-mode skips) on macOS M4 Max; 56/56 with the Metal skip lifted locally against a Quadrants build containing the issue #805 fix. Full suite also validated on linux CPU (fp64) and CUDA (fp32) with the tightened tolerances.
- Every strengthened test was verified to FAIL on current main and pass with this PR.
- The checkpoint-replay fix was validated by instrumenting the backward replay: replayed substep states now match the forward pass bit-exactly.

## Checklist:

- [x] I read the **CONTRIBUTING** document.
- [x] I followed the `Submitting Code Changes` section of **CONTRIBUTING** document.
- [x] I tagged the title correctly (including BUG FIX/FEATURE/MISC/BREAKING)
- [x] I updated the [documentation](https://github.com/Genesis-Embodied-AI/genesis-doc) accordingly or no change is needed.
- [x] I tested my changes and added instructions on how to test it for reviewers.
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.

### CODE REVIEW

File: genesis/engine/entities/rigid_entity/rigid_entity.py [hunk 1]
```diff
@@ -59,7 +59,12 @@ def wrapper(self, *args, **kwargs):
             bound = sig.bind(self, *args, **kwargs)
             bound.apply_defaults()
             args_dict = dict(tuple(bound.arguments.items())[1:])
-            self._update_tgt(fun.__name__, args_dict)
+            # Key the slot by (method, dofs subset) so same-step calls on distinct subsets (e.g. arm and gripper
+            # force control) each keep their own entry and gradient path; keyed by method alone, the second call
+            # would evict the first from the tape.
+            dofs_idx_local = args_dict.get("dofs_idx_local")
+            subset = None if dofs_idx_local is None else tuple(np.asarray(dofs_idx_local).reshape(-1).tolist())
```
[the remark this answers is not in the document]
Comment (duburcqa) [reply genesis/engine/entities/rigid_entity/rigid_entity.py]: fixed.

**Comment by duburcqa** [comment]
@codex review

File: genesis/engine/solvers/rigid/rigid_solver.py [hunk 2]
```diff
@@ -1347,40 +1343,66 @@ def _func_constraint_force(self):
...
-        kernel_accumulate_constraint_solver_grads(
-            self.dyn_state, self.constraint_solver.constraint_state, self.rigid_info, self.rigid_config
-        )
+        if gs.use_zerocopy:
+            force_grad = qd_to_torch(self.dyn_state.dofs.force.grad, copy=False)
+            dL_dforce = qd_to_torch(constraint_state.dL_dforce, copy=False)
+            force_grad.add_(dL_dforce)
+            mass_mat_grad = qd_to_torch(self.rigid_info.mass_mat.grad, copy=False)
+            dL_dM = qd_to_torch(constraint_state.dL_dM, copy=False)
+            mass_mat_grad.add_(dL_dM)
```
[the remark this answers is not in the document]
Comment (duburcqa) [reply genesis/engine/solvers/rigid/rigid_solver.py:1370]: fixed

**Comment by duburcqa** [comment]
@codex review
